[ASan] Allow for passing AddressSanitizer command line options through the AddressSanitizerOptions struct.

usama54321 opened the pull request:
This patch adds the ability to pass values for the command line options of -max-inline-poisoning-size, -instrumentation-with-calls-threshold and -asan-guard-against-version-mismatch through the AddressSanitizerOptions struct. The motivation is to use these new options when using the pass in Swift.

rdar://118470958

wrotki commented:
LGTM. I assume that existing tests like llvm/test/Instrumentation/AddressSanitizer/instrumentation-with-call-threshold.ll should cover this slightly changed functionality.

usama54321 commented:
lint fixes

vitalybuka commented on llvm/include/llvm/Transforms/Instrumentation/AddressSanitizer.h:
```diff
@@ -20,12 +20,20 @@ namespace llvm {
 class Module;
 class raw_ostream;
 
+static const int InstrumentationWithCallsThreshold = 7000;
```
Please move them into 

AddressSanitizerOptions

they do not deserve to be in llvm:: top namespace

vitalybuka commented on llvm/lib/Transforms/Instrumentation/AddressSanitizer.cpp:
```diff
@@ -644,18 +643,28 @@ namespace {
-                   bool UseAfterScope = false,
-                   AsanDetectStackUseAfterReturnMode UseAfterReturn =
-                       AsanDetectStackUseAfterReturnMode::Runtime)
+  AddressSanitizer(
```
May I ask you to remove default arguments from these constructor, including existing ones. In a separate PR is possible.

This is internal class and it's constructed just onces.
It will make code cleaner

vitalybuka commented on llvm/lib/Transforms/Instrumentation/AddressSanitizer.cpp:
```diff
@@ -782,9 +793,13 @@ class ModuleAddressSanitizer {
                          bool Recover = false, bool UseGlobalsGC = true,
```
same about default args for ModuleAddressSanitizer

usama54321 committed:
[ASan] Allow for passing AddressSanitizer command line options through
the AddressSanitizerOptions struct.

This patch adds the ability to pass values for the command line options
of -max-inline-poisoning-size, -instrumentation-with-calls-threshold and
-asan-guard-against-version-mismatch through the AddressSanitizerOptions
struct. The motivation is to use these new options when using the pass
in Swift.

rdar://118470958
